File: Binary Trees/236. Lowest_Common_Ancestor_of_a_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        
        if(root==null || root==p|| root==q)
        {
            return root;  //keeping in mind that the general order of the tree is a pre-order traversal where the node comes first
        }

        TreeNode left=lowestCommonAncestor(root.left,p,q);  //storing the value that comes from the left sub-tree
        TreeNode right=lowestCommonAncestor(root.right,p,q); //storing the value that comes from the right sub-tree 
    
        if(left==null)
        {
            return right;
        }
        else if(right==null)
        {
            return left;
        }
        else
        {
            return root;
        }

    }
}

File: Binary Trees/236. Lowest Common Ancestor of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == nullptr || root->val == p->val || root->val == q->val)
        return root;
 
    TreeNode * left_lca  = lowestCommonAncestor (root->left,  p, q); 
    TreeNode * right_lca = lowestCommonAncestor(root->right, p, q); 

    if (left_lca && right_lca)
        return root;

    if (left_lca == NULL)
        return right_lca;
    else
        return left_lca;
    }
};
